api/capi/src/nntrainer.cpp

```diff
@@ -483,6 +458,157 @@ int ml_train_optimizer_set_property(ml_train_optimizer_h optimizer, ...) {
   return status;
 }
 
+int ml_train_model_set_dataset(ml_train_model_h model,
+                               ml_train_dataset_h dataset) {
+  int status = ML_ERROR_NONE;
+  ml_train_model *nnmodel;
+  ml_train_dataset *nndataset;
+
+  ML_NNTRAINER_CHECK_MODEL_VALIDATION(nnmodel, model);
+  ML_NNTRAINER_CHECK_DATASET_VALIDATION(nndataset, dataset);
+
+  std::shared_ptr<nntrainer::NeuralNetwork> NN;
+  std::shared_ptr<nntrainer::DataBuffer> data;
+
+  NN = nnmodel->network;
+  data = nndataset->data_buffer;
+
+  returnable f = [&]() { return NN->setDataBuffer(data); };
+
+  status = nntrainer_exception_boundary(f);
+  if (status == ML_ERROR_NONE) {
+    nndataset->in_use = true;
+    if (nnmodel->dataset)
+      nnmodel->dataset->in_use = false;
+    nnmodel->dataset = nndataset;
```

Comment:
It seems that if there is any previous dateset, it could dangle although it would be user's responsibility.

Reply:
If there is any previous dataset, then there are two cases - 
1. Already set dataset was created using ini file. In this case, it is a shared_ptr internally and will be automatically freed.
2. Already set dataset was set by user. In this case, user has access to ml_train_dataset_h which must be appropriately freed by the user using ml_train_dataset_destroy.

Reply:
Thanks :)